### PR TITLE
reload library based on fw_image_flag in fw manifest

### PR DESCRIPTION
**Pull request by RanderWang**
Discussed in converged fw sync meeting and got a conclusion that use a manifest flag 
We will set fw_image_flag for library reload in fw config file so library will be reloaded by default. This will not affect ref fw.

Validated on mtl device, with mtl-006 and rimage PR https://github.com/thesofproject/rimage/pull/191

### CODE REVIEW

File: sound/soc/sof/ipc4-priv.h [hunk 1]
```diff
@@ -18,6 +18,18 @@
...
+	SOF_IPC4_ROM,
+	SOF_IPC4_BASEFW,
+	SOF_IPC4_LIBRARY
+};
+
+#define SOF_IPC4_RESERVED		BIT(0)
+#define SOF_IPC4_IMAGE_TYPE		GENMASK(2, 1) /* 0 - ROM, 1 - Base FW, 2 -library */
+#define SOF_IPC4_GET_IMAGE_TYPE(x)	(((x) & SOF_IPC4_IMAGE_TYPE) >> 1)
+#define SOF_IPC4_RELOCATABLE_LIB	BIT(3)
+#define SOF_IPC4_LIBRARY_RELOAD	BIT(4)
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-priv.h]: Please align the lines

File: sound/soc/sof/intel/hda-loader.c [hunk 2]
```diff
@@ -519,14 +519,16 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
...
 	struct firmware stripped_firmware;
 	struct sof_ipc4_msg msg = {};
 	struct snd_dma_buffer dmab;
 	int ret, ret1;
 
-	/* IMR booting will restore the libraries as well, skip the loading */
-	if (reload && hda->booted_from_imr)
+	/* if IMR booting is enabled and library reload is not defined, skip the loading */
+	if (reload && hda->booted_from_imr &&
+	    !(ipc4_data->fw_image_flags & SOF_IPC4_LIBRARY_RELOAD))
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-loader.c]: I would opt for a more generic name still, like `SOF_IPC4_CONTEXT_NOT_PRESERVED` or something (probably toss in _D3_ to _IMR_? or just use comment to clarify it's meaning and keeping it generic).
If we ever want or need to experiment with the context save/restore at large in other parts of the code.

**Review by ujfalusi — COMMENTED** [review]
@RanderWang, I did not even knew that the fw image have existing flag, this makes it clear. Great!
Few nitpicks from my side.

File: sound/soc/sof/ipc4-priv.h (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/ipc4-priv.h]: @mwasko Are we aligned on this?

Comment (RanderWang) [reply sound/soc/sof/ipc4-priv.h]: thanks, updated

File: sound/soc/sof/intel/hda-loader.c (hunk 2 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-loader.c]: @ujfalusi I use the name SOF_IPC4_FW_CONTEXT_LOST, how about it ? Thanks!

**Comment by RanderWang** [comment]
updated flag name

**Review by ujfalusi — APPROVED** [review]
@RanderWang, @mwasko, assuming we are aligned with SOF and reference firmware.

**Comment by RanderWang** [comment]
no need

**Pull request closed without merging**